src/VisualStudio/Next/Remote/ServiceHubRemoteHostClient.JsonRpcSession.cs

```diff
@@ -160,6 +154,62 @@ private class SnapshotJsonRpcClient : JsonRpcClient
                     }
                 }
 
+                private async Task WriteAssetAsync(ObjectWriter writer, byte[][] checksums)
+                {
+                    // special case
+                    if (checksums.Length == 0)
+                    {
+                        await WriteNoAssetAsync(writer).ConfigureAwait(false);
+                        return;
+                    }
+
+                    if (checksums.Length == 1)
+                    {
+                        await WriteOneAssetAsync(writer, checksums[0]).ConfigureAwait(false);
+                        return;
+                    }
+
+                    await WriteMultipleAssetsAsync(writer, checksums).ConfigureAwait(false);
+                }
+
+                private Task WriteNoAssetAsync(ObjectWriter writer)
+                {
+                    writer.WriteInt32(0);
+                    return SpecializedTasks.EmptyTask;
+                }
+
+                private async Task WriteOneAssetAsync(ObjectWriter writer, byte[] checksum)
+                {
+                    var service = ChecksumScope.Workspace.Services.GetRequiredService<ISolutionChecksumService>();
+
+                    var checksumObject = service.GetChecksumObject(new Checksum(checksum), _source.Token);
+                    writer.WriteInt32(1);
+
+                    writer.WriteValue(checksum);
+                    writer.WriteString(checksumObject.Kind);
+
+                    await checksumObject.WriteToAsync(writer, _source.Token).ConfigureAwait(false);
+                }
+
+                private async Task WriteMultipleAssetsAsync(ObjectWriter writer, byte[][] checksums)
+                {
+                    var service = ChecksumScope.Workspace.Services.GetRequiredService<ISolutionChecksumService>();
+
+                    var checksumObjectMap = service.GetChecksumObjects(checksums.Select(c => new Checksum(c)), _source.Token);
+                    writer.WriteInt32(checksumObjectMap.Count);
+
+                    foreach (var kv in checksumObjectMap)
+                    {
+                        var checksum = kv.Key;
+                        var checksumObject = kv.Value;
+
+                        writer.WriteValue(checksum.ToArray());
+                        writer.WriteString(checksumObject.Kind);
+
+                        await checksumObject.WriteToAsync(writer, _source.Token).ConfigureAwait(false);
```

Comment:
I guess I will rename the method to WriteObjectTo rather than WriteToAsync. since it write the object the checksum object is created from.